ML support (#5535)
mendix/docs: content/en/docs/refguide/modeling/integration/machine-learning-kit/using-ml-kit/_index.md

## 1 Introduction

Studio Pro [9.23](/releasenotes/studio-pro/9.23/) and above supports machine learning (ML) model integration in Mendix apps. Collectively called the *Machine Learning (ML) Kit*, this functionality allows Mendix developers to deploy a ML model built using common ML framework and language into Mendix Studio Pro runtime.

## 2 Usage {#usage}

To use a ML model in your app, you need to convert it to ONNX, then import it to create a model mapping.

### 2.1 Converting Your Model to ONNX {#convert-ml-model}

To embed your ML model into a Mendix app, you need to convert your model into the ONNX format. Depending upon the frameworks and tools used to create the ML model, there are many tools and sources to convert a model to ONNX format.

Examples include the following:

* [PyTorch](https://pytorch.org/docs/stable/onnx) 
* [TensorFlow to ONNX](https://github.com/onnx/tensorflow-onnx)
* [Scikit-learn to ONNX](https://github.com/onnx/sklearn-onnx)

### 2.2 Importing a Model and Creating the Mapping {#import-model}

> To use the ML model in your app, you will import it to create a mapping document.

Reviewer: ```suggestion
To use the ML model in your app, you import it to create a mapping document.
```

1. Right-click on the desired module and select **Add other > ML model mapping**.
2. Add a name for the mapping document.
3. Click **Import Model** to import your ONNX file.

> This will generate two non-persistable entities representing your ML Model input and outputs (see [Persistable and Non-Persistable Entities](#persist-nonpersist-entities) below).

Reviewer: ```suggestion
This generates two non-persistable entities representing your ML Model input and outputs (see [Persistable and Non-Persistable Entities](#persist-nonpersist-entities) below).
```

> This will generate two non-persistable entities representing your ML Model input and outputs (see [Persistable and Non-Persistable Entities](#persist-nonpersist-entities) below).

Reviewer: ```suggestion
This will generate two non-persistable entities representing your ML Model input and outputs (see the [Persistable and Non-Persistable Entities](#persist-nonpersist-entities) section below).
```

{{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/ml-model-created-entities.png" alt="Two non-persistable entities generated when importing an ONNX file." >}}

> If error CE1790 appears, you will need to [set dynamic tensor shapes](#set-dynamic-shapes).

Reviewer: ```suggestion
If error CE1790 appears, you need to [set dynamic tensor shapes](#set-dynamic-shapes).
```

#### 2.2.1 Locating the Model Directory {#model-directory}

> The ML model in ONNX format will be placed in the `mlsource/<your_module_name>` folder from the App/Show App Directory in Explorer.

Reviewer: Please see our style guide on using the Future Tense, we try to use Present Tense where possible.

This way, the files will be packaged with your app, and you can easily refer to them in microflow actions.
{{% /alert %}}

#### 2.2.2 Tensor Shapes {#tensor-shapes}

Tensors inject and retrieve information from a machine learning model. A distinctive trait of such an entity mapping is the tensor Shape, or the definition of the dimensions a tensor has.

##### 2.2.2.1 Static Shapes {#static-shapes}

Studio Pro detects models with static shapes automatically and display them in the mapping. The image below is a ResNet50 mapping with static dimensions that accepts 1 image of 3 channels (colors) with a size of 224x224 pixels:

{{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/mapping-static-dimensions.png" alt="ResNet50 mapping with static dimensions. Described in the paragraph above." >}}

##### 2.2.2.2 Dynamic Shapes {#dynamic-shapes}

Several models (including EasyOCR and Yolo) have tensors with dynamic shapes. In these cases, decide on a tensor mapped static shape and attribute type in a mapped entity before integrating in microflows.

> ###### 2.2.2.2.1 Setting Dynamic Tensor Shape Mapping (Error 1790) {#set-dynamic-shapes}

Reviewer: By "Error" do you mean consistency error? They are usually referred to as CE{error_number} in docs because this is how they appear in the UI for users.

When importing your model, you might encounter error CE1790. Go to the error and double-click on the affected mapping line to open the **Edit ML model input shape**s dialog box:

{{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/edit-model-input-shapes.png" alt="Edit ML model input shape dialog box." >}}

In this case, the -1 dimensions should be configured before using the mapping in a [Call ML model](/refguide/call-ml-model/) activity. Once filled, static tensor shapes of an output mapping will be automatically calculated based on configured dimensions of the input entity mappings:

{{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/completed-mapping.png" alt="Completed ML mapping. Described in the paragraph above." >}}

> {{% alert color="info" %}}Some these shapes should be handled in [Java Action pre- and post-processors](#java-pre-post) you may have.{{% /alert %}}

Reviewer: ```suggestion
{{% alert color="info" %}}Some of these shapes should be handled in [Java Action pre- and post-processors](#java-pre-post) you may have.{{% /alert %}}
```

#### 2.2.3 Persistable and Non-Persistable Entities {#persist-nonpersist-entities}

After importing a model, two [non-persistable](/refguide/persistability/) entities will be created using the input type.

For non-structured data, such as most of the tensors for Neural Networks, an entity attribute will be a `String`.  This is because Base64 is used to to encode the tensors to and from MLKit.

##### 2.2.3.1 Converting to Persistable Entities

You can convert these entities into [peristable](/refguide/persistability/) ones and use other types, such as **Binary**. This can be done to decrease inference latency of ML mappings document used in the [Call ML model](/refguide/call-ml-model/) activity. 

To convert an entity, do the following:

1. Double-click on the entity to open its **Properties**.
2. Select **Yes** in the **Persistable** section.

> You now have a [peristable](/refguide/persistability/) entity in your domain model that you can set as the type you need, and that can be used in [pre- and post-processors](#pre-post-processors).

Reviewer: ```suggestion
You now have a [persistable](/refguide/persistability/) entity in your domain model that you can set as the type you need, and that can be used in [pre- and post-processors](#pre-post-processors).
```

 
### 2.3 Using the Model in a Microflow {#use-model-microflow}

Once the ML [mapping document](#import-model) is created by importing the model, the ML model is available in the Studio Pro. 

> Drag and drop the [Call ML model](/refguide/call-ml-model/) activity from the toolbar in the Microflow editor to use call and use your ML model in your application logic..

Reviewer: ```suggestion
Drag and drop the [Call ML model](/refguide/call-ml-model/) activity from the Toolbox in the microflow editor to use call and use your ML model in your application logic.
```

#### 2.3.1 Example of a Model Microflow

In the microflow below, a [Create object](/refguide/create-object/) activity creates an object of input entity type, mapped to input tensors in the ML mapping document. The [Call ML model](/refguide/call-ml-model/) activity calls the model, and a [Change object](/refguide/change-object/) activity updates business object with predicted value.

{{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/drag-action-into-microflow.png" alt="Completed ML mapping. Described in the paragraph above." >}}

##### 2.3.1.2 Call ML Model Activity Details

> The **Call ML model** activity is an **ML Kit** activity in the **Toolbar**. Detailed documentation on the properties of this activity can be found in the [Call ML model](/refguide/call-ml-model/) page.

Reviewer: ```suggestion
The **Call ML model** activity is an **ML Kit** activity in the **Toolbox**. Detailed documentation on the properties of this activity can be found in the [Call ML model](/refguide/call-ml-model/) page.
```

> The **Call ML model** activity is an **ML Kit** activity in the **Toolbar**. Detailed documentation on the properties of this activity can be found in the [Call ML model](/refguide/call-ml-model/) page.

Reviewer: Please see the cross-reference style in our Style guide.

##### 2.4 Logistic Regression Example

The images below show an example of a logistic regression, its [netron.app](https://netron.app/) schema and the component display, along with the ML Mapping:

{{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/output-type.png" alt="Example of a logistic regression." >}}

> {{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/ml-model-created-entities.png" alt="Example of a netron.app schema." >}}

Reviewer: This is not the best practice to have several images one by one without a text. Can the description be added? Or can they be grouped in one image/diagram?

{{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/call-ml-mapping.png" alt="Example of the mapping call." >}}

> Create pre-processed inputs with **Create object** activity for inference inside your Mendix App, and consume the inference output with microflow [expressions](/refguide/expressions/).

Reviewer: ```suggestion
Create pre-processed inputs with **Create object** activity for inference inside your Mendix app, and consume the inference output with microflow [expressions](/refguide/expressions/).
```

> A **Create Object** activity is used for the first part in order to transfer entities from the Domain Model to the model input object. 

Reviewer: ```suggestion
A **Create Object** activity is used for the first part in order to transfer entities from the domain model to the model input object. 
```

> {{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/create-object.png" alt="Details of the log message in a microflow example." >}}

Reviewer: This is not the best practice to have several images one by one without a text. Can the description be added? Or can they be grouped in one image/diagram?

{{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/create-variable.png" alt="Details of the log message in a microflow example." >}}

{{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/change-object.png" alt="Details of the log message in a microflow example." >}}

## 3 Integrating Models with Pre-processors and Post-processors {#pre-post-processors}

> Integrating machine learning models can sometimes require using a more complex structure. This incudes having a pre-processor, the model itself, and a post-processor:

Reviewer: ```suggestion
Integrating machine learning models can sometimes require using a more complex structure. This includes having a pre-processor, the model itself, and a post-processor:
```

{{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/pre-post-processing-microflow.png" alt="Example of a microflow that includes a pre-processor and post-processor." >}}

For information on design patterns that include pre-processors and post-processors, see [Pre/Post-Processor Design Patterns](/refguide/machine-learning-kit/design-patterns/pre-post-processor-patterns/).

### 3.1 Processing {#processing}

> The *pre-processor* is the block of code that manipulates the data before being injected into the model. Some common examples of the pre-processing are [normalization](https://towardsdatascience.com/understand-data-normalization-in-machine-learning-8ff3062101f0) and [missing data handling](https://towardsdatascience.com/7-ways-to-handle-missing-values-in-machine-learning-1a6326adf79e). 

Reviewer: Please see our style guide for links and cross-references.

The required pre-processing varies a lot with each model and implementation, and can even be part of the ONNX file itself (see [Pre and post processing](https://onnxruntime.ai/docs/reference/build-web-app.html#pre-and-post-processing) in the ONNX runtime documentation). It is highly coupled with the model, its training process, and the domain of knowledge.

#### 3.1.1 Multidimensional Outputs

> Currently, it is required to have any multidimensional data as a string encoded using the base64 format. Any multidimensional data has to be encoded, either in string base64 format if you are using non-persistable entities, or binary streams if you opt for persistable entities for the model in the mapping (see [Persistable and Non-Persistable Entities](#persist-nonpersist-entities) in this document).

Reviewer: ```suggestion
Currently, it is required to have any multidimensional data as a string encoded using the base64 format. Any multidimensional data has to be encoded, either in string base64 format if you are using non-persistable entities, or binary streams if you opt for persistable entities for the model in the mapping (see the [Persistable and Non-Persistable Entities](#persist-nonpersist-entities) section in this document).
```

Below are two models (a Scikit-learn [random forest](https://www.kaggle.com/code/prashant111/random-forest-classifier-tutorial) and a ResNet50 network), and their [netron.app](https://netron.app/) schemas, and how these are rendered by the [Call ML model](/refguide/call-ml-model/) activity.

> {{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/advanced-model-rendering.png" alt="Example of how two models are rendered by the call ML model activity." >}}

Reviewer: Is this image a table? It is quite small and does present it in the best way. Please consider redoing it in Snagit.

> In this example, the [Call ML Model][Call ML model](/refguide/call-ml-model/) activity adapts to the model output. If you intend to use a multidimensional output in this stage (such as in the ResNet), it will come out string-encoded using base64 (link to section about persistable entities), and will probably require post-processing.

Reviewer: The link to the section is missing
```suggestion
In this example, the [Call ML Model][Call ML model](/refguide/call-ml-model/) activity adapts to the model output. If you intend to use a multidimensional output in this stage (such as in the ResNet), it will come out string-encoded using base64, and will probably require post-processing.
```

> Once you import your model, the Call ML Model activity allows you to add your input variable coming from the previous action, and name your output to be used in any ulterior steps:

Reviewer: ```suggestion
Once you import your model, the **Call ML Model** activity allows you to add your input variable coming from the previous action, and name your output to be used in any ulterior steps:
```

{{< figure src="/attachments/refguide/modeling/integration/ml-kit/ml-kit/name-your-output.png" alt="Example of naming the object output." >}}

### 3.2 Pre-processing and Post-processing using Java Actions {#java-pre-post}

> Data transformations are usually complex tasks, and often require mathematical libraries or even more complex pieces of software (think OpenCV for computer vision). As a result, sometimes they are not integrated into the model. In this case, the best way to perform these transformations in Mendix using Java Actions (see [Extending Your Application with Custom Java](/refguide/extending-your-application-with-custom-java/)).

Reviewer: ```suggestion
Data transformations are usually complex tasks, and often require mathematical libraries or even more complex pieces of software (think OpenCV for computer vision). As a result, sometimes they are not integrated into the model. In this case, the best way to perform these transformations in Mendix is using Java Actions (see [Extending Your Application with Custom Java](/refguide/extending-your-application-with-custom-java/)).
```

 
Read the sections below for an example of pre-processing and post-processing using Java Actions in Mendix.

> #### 3.2.1 Pre-processing using Java Actions

Reviewer: ```suggestion
#### 3.2.1 Pre-processing Using Java Actions
```

> This uses a [Random Forest](https://www.kaggle.com/code/prashant111/random-forest-classifier-tutorial) example trained with the Iris/Setosa dataset. The output is a long value, representing a flower type. The input is a base64 encoded float array, as we use Base64 encoded strings for passing multidimensional data to the Call ML Model action at this stage of development.

Reviewer: "This" what? section?

```
    @java.lang.Override
    public IMendixObject executeAction() throws Exception 
    {
      this.Parameter = this.__Parameter == null ? null : iris_randomforest.proxies.Iris.initialize(getContext(), __Parameter);
    
      // BEGIN USER CODE
      // 1. convert entity object to float array
      float[] inputFeature = new float[4];
      inputFeature[0] = Parameter.getsepalLength().floatValue();
      inputFeature[1] = Parameter.getsepalWidth().floatValue();
      inputFeature[2] = Parameter.getpetalLength().floatValue();
      inputFeature[3] = Parameter.getpetalWidth().floatValue();
      // 2. convert input feature array to base 64
      InputStream is = MLKit.toInputStream(inputFeature);
      String base64 = MLKit.toBase64(is);
      // 3. create output entity object
      IMendixObject outputObject = Core.instantiate(getContext(), "Iris_RandomForest.ML_Input_Entity_Iris_RF_ML_Model");
      outputObject.setValue(getContext(), "Float_input", base64);
      return outputObject;
      // END USER CODE
    }
```

#### 3.2.2 Post-processing using Java Actions

> The rationale for using a post-processor is quite similar to the pre-processor but in the opposite direction. We’re going to extract something out of the model and do something with it. For more complex models, the output can be multidimensional, and then decoding is required. In this example, the post-processor is to map it using a class map in order to get not ids nor probabilities, but the class name, using a file already present in the artifacts directory.

Reviewer: "We’re going to extract something out of the model and do something with it."
this phase is very unclear to me. We also avoid using "we" because it usually means "we, Mendix". Here it means "you" - you, as a user, are going to do something. I would rewrite this sentence and make it clear what a user is going to do.

```
    @java.lang.Override
    public java.lang.String executeAction() throws Exception
    {
    this.Parameter = this.__Parameter == null ? null : iris_randomforest.proxies.ML_Output_Entity_Iris_RF_ML_Model.initialize(getContext(), __Parameter);
    
      // BEGIN USER CODE
      if (Parameter.getOutput_label() == 0) {
        return "Setosa";
      } else if (Parameter.getOutput_label() == 1) {
        return "Versicolour";
      } else
        return "Virginica";
      // END USER CODE
    }
```

For detailed information on pre-processing and post-processing patterns of machine learning designs, see [Pre/Post-Processor Design Patterns](/refguide/machine-learning-kit/design-patterns/pre-post-processor-patterns/).

## 4 Supported Frameworks and Libraries {#supported-frameworks}

To ensure that your model will work with the [Call ML Model](/refguide/call-ml-model/) activity, follow the guidelines in this section.

> To learn more about design patterns that can be applied to the Inference process, and specifications that for pre/post-processing while using Mendix components, Java Actions, or embedding it into the model directly, see our documentation on [Design Patterns](/refguide/machine-learning-kit/design-patterns/).

Reviewer: ```suggestion
To learn more about design patterns that can be applied to the Inference process, and specifications that for pre/post-processing while using Mendix components, Java actions, or embedding it into the model directly, see our documentation on [Design Patterns](/refguide/machine-learning-kit/design-patterns/).
```

### 4.1 Model Building

ML Kit comes with native support to [ONNX Runtime](https://onnxruntime.ai/about.html), which means Mendix developers are free to use the right machine learning framework or library for their use cases. Mendix developers can use the native ONNX operators or use the major machine learning libraries like Tensorflow, PyTorch, scikit-learn, and XGBoost to develop their models, preprocessing, and post-processing actions.

Artifacts (including HDF5, protocol buffers, and pickle) should be converted to ONNX computation graphs in order to integrate them into Mendix. 

The ONNX community develops and maintains the converters for many ML frameworks. Check out the [ONNX repository](https://github.com/onnx/onnxmltools) and [supported tools](https://onnx.ai/supported-tools.html) for further information.

### 4.2 Pre-Trained Models

To integrate a pre-built (propriety or public) ML model into a Mendix app, consider the following:

* Compatibility of the model internal representation (IR) version with the ML Kit
* Compatibility of the model OpSets with the ML Kit

> See [Versioning] for details on versioning schemes

Reviewer: For details on versioning schemes, see the [Versioning](#versioning) section below.

### 4.3 Versioning {#versioning}

Use [ONNX versioning schemes](https://github.com/onnx/onnx/blob/main/docs/Versioning.md) while creating and integrating ONNX computation graphs with Mendix apps. Mendix Studio Pro guides developers to comply with the suggested internal representation, and package OpSets.

In case of version conflicts in models (or pre- / post-processors), try the [version converter tool](https://github.com/onnx/onnx/blob/main/docs/VersionConverter.md) to align with the suggested IR and OpSets.

The released versions of the ML Kit and supported ONNX Runtime versions can be found below.

| Mendix Studio Pro | ML Kit   | ONNX Runtime |
| ----------------- | -------- | ------------ |
| 9.23              | Public Beta | 1.11.0       |

## 5 Monitoring and Troubleshooting

Machine learning models might be placed in the core of a Mendix application that makes automated decisions, and data quality is a crucial part of a machine learning system. Because of this, Mendix developers need to implement and configure monitoring tools that track not only service health but also the input and output of the MLKit actions.

> The [Monitoring and Troubleshooting](/howto/monitoring-troubleshooting/) guide in Studio Pro helps

Reviewer: Please follow our style guide for cross-references.

> Typically microflows that contain an MLKit activity may also be composed of various helper activities (e.g. Java Actions, Aggregate List Activities) in order to provide necessary input features to the model or parse its outputs. Tools that are provided by Studio Pro can be easily used for monitoring and troubleshooting. 

Reviewer: ```suggestion
Typically microflows that contain an MLKit activity may also be composed of various helper activities (e.g. Java actions, Aggregate List activities) in order to provide necessary input features to the model or parse its outputs. Tools that are provided by Studio Pro can be easily used for monitoring and troubleshooting. 
```

Besides the traditional Mendix monitoring and troubleshooting tools, we share some tips for Mendix developers to get insights about inference calls in the development and production phases. 

One might need to observe the inputs/outputs of a model that is used in an MLKit activity, assess the performance (latency) of making a prediction, or debug an individual pre/post-processing unit on a test sample to verify that the Microflow works and to catch potential bugs in development time before rolling out the application to the production environments. The StudioPro allows us to fulfill these aims in many different ways.

### 5.1 Troubleshooting

The [Errors Pane](/refguide/errors-pane/) in Studio Pro shares informative messages about the metadata of the models that are consumed by the **Call ML Model** activity. Warning messages do not block deployment, but the application may not work seamlessly. All error messages have to be resolved in order to execute applications in local environments or deploy applications into the production environments.

>     Warning

Reviewer: We usually do not document warnings. Do error codes have these links in the Errosr Pane?

- CE9997: ONNX IR version Incompatibility. See [ONNX versioning docs](https://github.com/onnx/onnx/blob/main/docs/Versioning.md) and version compatibility matrixes in the following section.
- CE9998: ONNX Opset Incompatibility. See [ONNX versioning docs](https://github.com/onnx/onnx/blob/main/docs/Versioning.md) and version compatibility matrixes in the following section.

>     Error

Reviewer: If the description of the error duplicates what is written in the UI I would remove it for maintenance pusposes. We usually add extra guidance when we document errors.

- CE9999: Missing input. The required model input has to be provided.

### 5.2 Console

The [Console](/refguide/view-menu/#console) pane displays the output of the Mendix [Runtime](/refguide/runtime/) while running an application.

#### 5.2.1 Mendix Runtime Logs

> The responsibility of the Mendix Runtime is to make an inference call on the model which is integrated into an app via the `Call ML Model` activity. While making a prediction on the model, Mendix Runtime publishes several metrics on the ML Engine log node available in the Console and application logs. These instance-based metrics are provided in each inference call to provide insights to the Mendix developers.

Reviewer: ```suggestion
The responsibility of the Mendix Runtime is to make an inference call on the model which is integrated into an app via the **Call ML Model** activity. While making a prediction on the model, Mendix Runtime publishes several metrics on the ML Engine log node available in the Console and application logs. These instance-based metrics are provided in each inference call to provide insights to the Mendix developers.
```
